### PR TITLE
feat(pattern): Handle empty alternatives

### PR DESCRIPTION
**Pull request by loewenheim**
This introduces a new `Optional` variant of `Token` that _optionally_ matches the contained list of tokens. These optional tokens are parsed from alternations containing empty branches, e.g. `{foo,bar,}` (note the trailing comma).

### CODE REVIEW

File: relay-pattern/src/wildmatch.rs [hunk 1]
```diff
@@ -109,6 +109,27 @@ where
...
+                            is_match_impl::<_, M>(h_current, &tokens)
+                        }),
+                    };
+                    // The brace match already matches to the end, if it is successful we can end right here.
+                    if matches {
+                        return true;
+                    }
+                    // No match, allow for backtracking.
+                    false
+                }
```
Comment (Dav1dde) [inline relay-pattern/src/wildmatch.rs]:
```suggestion
                Token::Optional(optional) => {
                    let optional = tokens.with_alternate(t_next, std::slice::from_ref(optional));
                    if is_match_impl::<_, M>(h_current, &optional) {
                        // There is a match with the optional token, we're done.
                        return true;
                    }
                    // Continue on without the optional token.
                    true
                }
```

Comment (loewenheim) [reply relay-pattern/src/wildmatch.rs]: Yes, that is _much_ better. Part of why I made it so complicated is not seeing the forest for the trees, but another part is also the fact that I thought we wouldn't be able to just call `with_alternate` with the optional token if it contains an `Alternate` token, and to be honest I still don't quite understand why that works.

File: relay-pattern/src/lib.rs [hunk 2]
```diff
@@ -1427,6 +1462,9 @@ mod tests {
         assert_pattern!("{foo,bar}", NOT "BAR");
         assert_pattern!("{foo,bar}", NOT "fooo");
         assert_pattern!("{foo,bar}", NOT "baar");
+        assert_pattern!("{foo,bar,}baz", "foobaz");
+        assert_pattern!("{foo,bar,}baz", "barbaz");
+        assert_pattern!("{foo,bar,}baz", "baz");
```
Comment (jjbayer) [inline relay-pattern/src/lib.rs:1467]: Should we add tests for `{foo,,bar}baz` and `{,foo,bar}baz`.

File: relay-pattern/src/lib.rs [hunk 3]
```diff
@@ -731,6 +760,12 @@ enum Token {
     Class { negated: bool, ranges: Ranges },
     /// A list of nested alternate tokens `{a,b}`.
     Alternates(Vec<Tokens>),
+    /// A list of optional tokens.
+    ///
+    /// This has no syntax of its own, it's parsed
+    /// from alternatives containing empty branches
+    /// like `{a,b,}`.
+    Optional(Vec<Token>),
```
Comment (Dav1dde) [inline relay-pattern/src/lib.rs]:
```suggestion
    Optional(Tokens),
```

Tokens gives additional guarantees, we should make sure the tokens contained here uphold the same guarantees.

File: relay-pattern/src/lib.rs (hunk 2 above)
Comment (loewenheim) [reply relay-pattern/src/lib.rs:1467]: Glad you mentioned this because `{,foo,bar}` currently fails.

Comment (Dav1dde) [reply relay-pattern/src/lib.rs:1467]: Should have a test for all added branches:

- `{foo*bar}`
- `{foo*bar,}`
- `{foo*bar,baz,}`
- `{foo*bar,baz}`
- `{,,,}`

Some of them probably already exist.

File: CHANGELOG.md [hunk 4]
```diff
@@ -11,6 +11,7 @@
 **Features**:
 
 - Implement zstd http encoding for Relay to Relay communication. ([#4266](https://github.com/getsentry/relay/pull/4266))
+- relay-pattern: Support empty branches in alternations. ([#4283](https://github.com/getsentry/relay/pull/4283))
```
Comment (Dav1dde) [inline CHANGELOG.md]:
```suggestion
- Support empty branches in Pattern alternations. ([#4283](https://github.com/getsentry/relay/pull/4283))
```

**Review by Dav1dde — APPROVED** [review]

**Pull request merged**